File: deps/Behavior/test/btunittest/Agent/CustomPropertyAgent.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "CustomPropertyAgent.h"

BEGIN_PROPERTIES_DESCRIPTION(UnityEngine::Vector3)
{
    REGISTER_PROPERTY(x);
    REGISTER_PROPERTY(y);
    REGISTER_PROPERTY(z);
}
END_PROPERTIES_DESCRIPTION()

int CustomPropertyAgent::IntMemberConst = 0;

CustomPropertyAgent::CustomPropertyAgent()
{
    init();
}

CustomPropertyAgent::~CustomPropertyAgent()
{
}

BEGIN_PROPERTIES_DESCRIPTION(CustomPropertyAgent)
{
    //CLASS_DISPLAYNAME(L"CustomPropertyAgent DisplayName")
    //CLASS_DESC(L"CustomPropertyAgent Desc")
    REGISTER_PROPERTY(IntProperty);
    REGISTER_PROPERTY(FloatPropertyReadonly);
    REGISTER_PROPERTY(BoolMemberReadonly);
    REGISTER_PROPERTY(IntMemberConst);
    REGISTER_PROPERTY(StringMemberReadonly);
    REGISTER_PROPERTY(Location);

    REGISTER_METHOD(FnWithOutParam);
}
END_PROPERTIES_DESCRIPTION()

File: deps/Behavior/integration/unity/Assets/Scripts/behaviac/BehaviacUnitTest/CustomPropertyAgent.cs
using UnityEngine;

[behaviac.TypeMetaInfo()]
public class CustomPropertyAgent : behaviac.Agent
{
    public void resetProperties() {
        this.Variables.Clear(false);
    }

    public void init() {
        base.Init();
        resetProperties();
    }

    public void finl() {
    }

    public int _int_member = 0;

    [behaviac.MemberMetaInfo("", "")]
    public int IntProperty {
        get
        {
            return _int_member;
        }
        set
        {
            _int_member = value;
        }
    }

    public float _float_member = 0;

    [behaviac.MemberMetaInfo()]
    public float FloatPropertyReadonly {
        get
        {
            return _float_member;
        }
    }

    [behaviac.MemberMetaInfo()]
    public readonly bool BoolMemberReadonly = false;

    [behaviac.MemberMetaInfo()]
    public const int IntMemberConst = 0;

    [behaviac.MemberMetaInfo(true)]
    public string StringMemberReadonly = "read only sting";

    [behaviac.MemberMetaInfo()]
    public Vector3 Location = new Vector3();

    [behaviac.MethodMetaInfo()]
    public void FnWithOutParam(out int param) {
        param = 1;
    }
}
